core/Segment.php

```diff
@@ -433,7 +433,7 @@ protected function getCleanedExpression($expression)
         // This is required to ensure segments like actionUrl!@value really do not include any visit having an action containing `value`
         if ($this->doesSegmentNeedSubquery($matchType, $name)) {
             $operator = $this->getInvertedOperatorForSubQuery($matchType);
-            $stringSegment = $name . $operator . $value;
+            $stringSegment = $name . $operator . preg_replace('/((?<!\\\)[,;])/', '\\\$1', $value);
```

Comment:
There are other variants that at least pass the added tests.

Encoding only the value:

```suggestion
            $stringSegment = $name . $operator . urlencode($value);
```

Encoding the complete segment:

```suggestion
            $stringSegment = urlencode($name . $operator . $value);
```

Does anyone find a case for one of the three solutions to be the "best" here?

Reply:
I think we should only either escape or encode the value only, not the whole segment. It's easier to debug/work with in the code when you can see the 'control'/functionally-significant characters, in this case the `,` and the `;` that we use as operators for chaining the segments.
And from the two options the `urlencode` is easier to comprehend vs looking at the preg_replace regex, so it would be my choice.

⚡ Code for future readability, not for current cool factor.

Reply:
Personally I think that encoding the whole segments feels wrong. I actually wonder why that works. Wouldn't expect that, but guess it's caused by the weird logic around trying to urldecode stuff to see if it "improves" the results.
Using `urlencode` on the value might match how the stuff is handled in other places, but as the segment and it's value should be decoded at that point already, it might imho be more correct to escape the problematic chars only.
At some point I think we might need to find a way to get rid of the double and tripple urlencoding of segments, and work with proper escaping instead, as it already causes various trouble. Adding another urlencode at this point, doesn't feel future-oriented to me.

But we could maybe move the `preg_replace` to it's own method called like `escapeSegmentValue` or similar

Reply:
Moved the scaping to a seperate function as mentioned (with a description of what is done to turn down the magic factor). We can always update the chosen way once we have information if any other option behaves differently/better.